### ISSUE TITLE
`bolt file upload` with directory as destination renames the file

### ISSUE THREAD

**Issue by DavidS** [puppetlabs/bolt#1930]
## Steps to Reproduce
```
david@zion:~/git/puppetlabs-facter_task (master)$ bundle exec bolt file upload --inventoryfile ./inventory.yaml -t localhost:2223 --debug ~/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz /root 
Loaded configuration from: ''
/home/david/gems/ruby/2.7.0/gems/puppet-6.16.0/lib/puppet/indirector.rb:49: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/david/gems/ruby/2.7.0/gems/puppet-6.16.0/lib/puppet/indirector/indirection.rb:95: warning: The called method `initialize' is defined here
/home/david/gems/ruby/2.7.0/gems/puppet-6.16.0/lib/puppet/util.rb:463: warning: URI.escape is obsolete
Loading modules from /home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/bolt-modules:/home/david/.puppetlabs/bolt/modules:/home/david/.puppetlabs/bolt/site-modules:/home/david/.puppetlabs/bolt/site:/home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/modules
Bolt may be installed as a gem. To use Bolt reliably and with all of its
dependencies, uninstall the 'bolt' gem and install Bolt as a package:
https://puppet.com/docs/bolt/latest/bolt_installing.html

If you meant to install Bolt as a gem and want to disable this warning,
set the BOLT_GEM environment variable.
Loaded inventory from /home/david/git/puppetlabs-facter_task/inventory.yaml
/home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/lib/bolt/cli.rb:335: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/lib/bolt/analytics.rb:117: warning: The called method `screen_view' is defined here
Started with 100 max thread(s)
Submitting analytics: {
  "v": 1,
  "cid": "da9fcaa3-fc43-47b2-a8be-368acd100cf9",
  "tid": "UA-120367942-1",
  "an": "bolt",
  "av": "2.15.0",
  "aip": true,
  "ul": "en-US",
  "cd1": "Debian bullseye/sid",
  "t": "screenview",
  "cd": "file_upload",
  "cd5": "human",
  "cd11": "user",
  "cd4": 1,
  "cd2": 2,
  "cd3": 4,
  "cd10": 2
}
/home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/lib/bolt/outputter/logger.rb:16: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/lib/bolt/outputter/logger.rb:26: warning: The called method `log_step_start' is defined here
Starting: file upload from /home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz to /root on localhost:2223
Completed analytics submission
Submitting analytics: {
  "v": 1,
  "cid": "da9fcaa3-fc43-47b2-a8be-368acd100cf9",
  "tid": "UA-120367942-1",
  "an": "bolt",
  "av": "2.15.0",
  "aip": true,
  "ul": "en-US",
  "cd1": "Debian bullseye/sid",
  "t": "event",
  "ec": "Transport",
  "ea": "initialize",
  "el": "ssh",
  "ev": 1
}
Uploading file /home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz to /root on ["localhost:2223"]
Uploading: '/home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz' to /root on localhost:2223
Started on localhost:2223...
Initializing ssh connection to localhost:2223
Completed analytics submission
Opened session
Executing: mkdir -m 700 /tmp/c5c8b314-6fe6-4694-8bd8-ec680b4b5193
Command returned successfully
Uploading /home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz, to /tmp/c5c8b314-6fe6-4694-8bd8-ec680b4b5193/root
Executing: mv -f /tmp/c5c8b314-6fe6-4694-8bd8-ec680b4b5193/root /root
Command returned successfully
Executing: rm -rf /tmp/c5c8b314-6fe6-4694-8bd8-ec680b4b5193
Command returned successfully
Closed session
{"target":"localhost:2223","action":"upload","object":"/home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz","status":"success","value":{"_output":"Uploaded '/home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz' to 'localhost:/root'"}}
Finished on localhost:2223:
  Uploaded '/home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz' to 'localhost:/root'
/home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/lib/bolt/outputter/logger.rb:18: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/david/gems/ruby/2.7.0/gems/bolt-2.15.0/lib/bolt/outputter/logger.rb:35: warning: The called method `log_step_finish' is defined here
Finished: file upload from /home/david/Downloads/puppet-enterprise-2019.2.1-el-8-x86_64.tar.gz to /root with 0 failures in 9.47 sec
Successful on 1 target: localhost:2223
Ran on 1 target in 9.47 sec
david@zion:~/git/puppetlabs-facter_task (master)$ 
```

The debug output shows that the file is uploaded to `/root/root` and this can be confirmed on the target machine.

## Expected Behavior
I expected to have the file uploaded to the `/root` directory, having the same filename as the local file.

When encountering this issue it would also have been helpful if bolt communicated the final location of the file in regular output without having to go into `--debug` mode.

## Environment
 - Version 2.15.0
 - Platform Debian bullseye/sid

## Additional Context
This is likely related to #1928.

**Comment by nicklewis** [comment]
Hmm this is interesting. We upload a tempfile which we then move into place. We calculate the tempfile name using the basename of the destination instead of the basename of the source, so it winds up renamed in this case. I _think_ we can upload it with basename based on the source and then move it into place and this should just work. If the destination is a directory, the file will be placed inside that directory with its original name and if the destination is a file or non-existent file it will be created with that name instead. I believe that fits the intuitive behavior.

```
$ bolt file upload -t target source_file.txt /dir
# This should create /dir/source_file.txt
$ bole file upload -t target source_file.txt /dir/dest.txt
# This should create /dir/dest.txt
```

We should check for similar issues with the other transports.

**Comment by DavidS** [comment]
That sounds sensible to me.

**Comment by lucywyman** [comment]
This was solved in #1945 and released in 2.16.0 - I just verified the fix, but let us know if you run into any other issues!
